Add test for improperly signed message

rargulati opened the pull request:
This PR introduces a test where an adversarial peer can sign a message with a key they
didn't originally register with. 

First, an adversarial peer will allow the improperly signed message to pass through 
validation (as they turn off strict verification, putting themselves at risk!). 

Next, it's tested that an honest peer, with strict verification on, will never see the 
malicious message, but they will see the correctly signed message.

rargulati committed:
Add test for improperly signed message

Test the case where an adversarial peer signs a message with a key they
didn't originally register with. First, we test that an adversarial peer
will allow the message to pass through validation as they turn off
strict verification (putting themselves at risk), but an honest peer
with strict verification on will never see the message!

rargulati committed:
Reduce time to wait

2 seconds per wait seemed a bit aggressive. Halved the total time of the test.

rargulati committed:
Ensure that the received message is the correct one

rargulati committed:
adversarial -> adversary

rargulati committed:
adversaryPeer -> adversary

rargulati committed:
Remove commented out code

bhost import was commented out.

pdyraga commented on floodsub_test.go:
```diff
@@ -943,3 +947,120 @@ func TestWithSigning(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	time.Sleep(time.Millisecond * 10)
```
Is it enough? I see we usually have `50` or `100` in this file.

pdyraga commented on floodsub_test.go:
```diff
@@ -943,3 +947,120 @@ func TestWithSigning(t *testing.T) {
+
+	if len(adversaryMessages) != 2 {
+		t.Fatalf("got %d messages, expected 2", len(adversaryMessages))
+	}
```
This is to make sure that the incorrect message is indeed published? I like this approach! Can we add a comment about the purpose of this code?

pdyraga commented on floodsub_test.go:
```diff
@@ -943,3 +947,120 @@ func TestWithSigning(t *testing.T) {
 	}
 }
+
+func TestImproperlySignedMessageNotRelayed(t *testing.T) {
```
We create two hosts so we have the following topology:
```
adversary <-> honest
```
and they are subscribed to the same topic named `"foobar"`.

In that case, I think this test is only about rejecting a message by `honest` peer. It's not about rejecting relayed message so it should be named `TestImproperlySignedMessageRejected`.

I also think we need another test if we want to have our network security requirements/assumptions covered. We need the following network topology:
```
adversary <-> relay_peer <-> destination_peer
```
- `adversary` should be subscribed to topic `"foobar"`
- `destination_peer` should be subscribed to topic `"foobar"`
- `relay_peer` should NOT be subscribed to any topic
- `adversary` sends a correct message to `destination_peer` and we need to make sure it's received
- `adversary` sends an incorrect message to `destination_peer` and we need to make sure it's rejected by `relay_peer`.

rargulati committed:
Change test name to reflect no relaying

Improperly named test. In this test, we're only testing that we reject
malicious messages.

rargulati committed:
Remove erroneous subscribe

rargulati committed:
Bump wait from 10 -> 50ms

This matches the waits in the rest of the package.

rargulati committed:
Add a note why we're expecting two messages

As the adversary has turned of signature verification, we should expect
to see two messages (the correctly signed and the incorrectly signed
one) when we publish. Those that have verification on, should only see
one.

rargulati committed:
Use names rather than index

We've aliased array elements, use the explicit names instead.

rargulati committed:
Better comment for signing with a different key

Note that the second message is the incorrectly signed message.
